distributed/shuffle/_rechunk.py

```diff
@@ -166,45 +169,310 @@ def rechunk_unpack(
         )
 
 
-def rechunk_p2p(x: da.Array, chunks: ChunkedAxes) -> da.Array:
-    import numpy as np
+class _Partial(NamedTuple):
+    """Information used to perform a partial rechunk along a single axis"""
+
+    #: Slice of the old chunks along this axis that belong to the partial
+    old: slice
+    #: Slice of the new chunks along this axis that belong to the partial
+    new: slice
+    #: Index of the first value of the left-most old chunk along this axis
+    #: to include in this partial. Everything left to this index belongs to
+    #: the previous partial.
+    left_start: int
+    #: Index of the first value of the right-most old chunk along this axis
+    #: to exclude from this partial.
+    #: This corresponds to `left_start` of the subsequent partial.
+    right_stop: int
+
+
+class _NDPartial(NamedTuple):
+    """Information used to perform a partial rechunk along all axes"""
+
+    #: n-dimensional slice of the old chunks along each axis that belong to the partial
+    old: NDSlice
+    #: n-dimensional slice of the new chunks along each axis that belong to the partial
+    new: NDSlice
+    #: Indices of the first value of the left-most old chunk along each axis
+    #: to include in this partial. Everything left to this index belongs to
```

Comment:
I'll add a few more unit tests around the chunking/partial logic to make this more explicit, this is not straightforward. Going over this piece of code again, I _think_ I can remove this bit of complexity by restructuring some of the logic. So far I've put this off as premature optimization, but it looks like it would also reduce the cognitive load.